### PR TITLE
fix(typescript): fix a memory leak (#322)

### PR DESCRIPTION
**Pull request by VitorLuizC**
## Rollup Plugin Name: `typescript`

This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [ ] yes (_bugfixes and features will not be merged without tests_)
- [x] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

If yes, then include "BREAKING CHANGES:" in the first commit message body, followed by a description of what is breaking. 

List any relevant issue numbers:

rollup/plugins#322

### Description

Just check if program already exists before creating a new one.

### CODE REVIEW

File: packages/typescript/src/index.ts [hunk 1]
```diff
@@ -34,14 +34,17 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
...
-        formatHost,
-        resolveModule,
-        parsedOptions,
-        writeFile(fileName, data) {
-          emittedFiles.set(fileName, data);
-        }
-      });
+      // Fixes a memory leak https://github.com/rollup/plugins/issues/322
+      if (!program) {
+        program = createWatchProgram(ts, this, {
```
Comment (shellscape) [inline packages/typescript/src/index.ts:39]: @NotWoods how can we best add a test for this to protect against regressions?

Comment (NotWoods) [reply packages/typescript/src/index.ts:39]: Unfortunately I'm not sure how to test for memory leaks short of re-running the test many times (which would have a large impact on our CI times). I think we should merge this in for now without an added test.

Comment (shellscape) [reply packages/typescript/src/index.ts:39]: right on.

**Comment by shellscape** [comment]
ping @NotWoods think this is good to go?

**Comment by shellscape** [comment]
thanks!

**Comment by finico** [comment]
@VitorLuizC @shellscape @NotWoods This change fixed the memory leaks, but led to a race between typescript's and rollup's watchers.
Now when saving the file, sometimes the bundle builds by rollup before typescript compiles the the file.
It's because independent watchers are used to track files and they don't wait for each other.

**Comment by shellscape** [comment]
@finico could you open that as a new issue with a reproduction?

**Comment by VitorLuizC** [comment]
@shellscape @NotWoods

I can make a new PR that closes current program and then open a new one, and it'll fix both issues.

```js
    buildStart() {
      emitParsedOptionsErrors(ts, this, parsedOptions);

      if (program) {
        program.close();
      }

      program = createWatchProgram(ts, this, {
          formatHost,
          resolveModule,
          parsedOptions,
          writeFile(fileName, data) {
            emittedFiles.set(fileName, data);
          }
        });
    }
```

**Comment by finico** [comment]
@VitorLuizC It has no sense to use WatchProgram that way - unnecessary overhead (about 700ms on my machine) over creation a new instance

**Comment by shellscape** [comment]
'Gents. It's best to open this on a _new issue_ for tracking. A reproduction is also essential because tests can be created from one. We let this PR in without tests due to it's nature, but I'm going to insist that any further PRs that modify how `program` is used be accompanied by a test. That is also something I would rather see discussed on a new issue. This PR is merged and closed, please move to a new issue.

**Comment by milky2028** [comment]
I'm seeing the race condition issue mentioned by @finico. I created an issue #418.

**Pull request merged**